File: backend/app/schemas/question_extraction.py
from __future__ import annotations

import uuid
from datetime import datetime
from decimal import Decimal
from typing import Annotated, Literal

from pydantic import BaseModel, ConfigDict, Field

from app.core.enums import QuestionExtractionRunStatus


class StrictQuestionExtractionSchema(BaseModel):
    """Strict public transport base for question extraction responses."""

    model_config = ConfigDict(extra="forbid", from_attributes=True)


class QuestionExtractionRunRead(StrictQuestionExtractionSchema):
    id: uuid.UUID
    source_document_id: uuid.UUID
    run_number: int
    status: QuestionExtractionRunStatus
    requested_by_user_id: uuid.UUID | None
    started_at: datetime | None
    completed_at: datetime | None
    failure_message: str | None


class QuestionCandidateRead(StrictQuestionExtractionSchema):
    id: uuid.UUID
    sequence_number: int
    extracted_text: str
    confidence: Decimal | None
    source_document_page_id: uuid.UUID | None
    page_number: int | None


class QuestionExtractionAnalysisPageRead(StrictQuestionExtractionSchema):
    source_document_page_id: uuid.UUID
    page_number: int


class QuestionExtractionTextSegmentRead(StrictQuestionExtractionSchema):
    type: Literal["text"]
    text: str


class QuestionExtractionMathSegmentRead(StrictQuestionExtractionSchema):
    type: Literal["math"]
    latex: str
    source_text: str
    display_mode: bool


QuestionExtractionContentSegmentRead = Annotated[
    QuestionExtractionTextSegmentRead | QuestionExtractionMathSegmentRead,
    Field(discriminator="type"),
]


class QuestionExtractionStructuredContentRead(StrictQuestionExtractionSchema):
    format_version: Literal[1]
    segments: list[QuestionExtractionContentSegmentRead] = Field(min_length=1)


class QuestionExtractionAnalysisOptionRead(StrictQuestionExtractionSchema):
    label: str | None
    text: str
    content: QuestionExtractionStructuredContentRead | None = None


class QuestionExtractionAnalysisCorrectionRead(StrictQuestionExtractionSchema):
    original_value: str
    normalized_value: str
    reason: str


class QuestionExtractionAnalysisQuestionRead(StrictQuestionExtractionSchema):
    id: uuid.UUID
    sequence_number: int
    question_number: str | None
    variant: str | None
    source_pages: list[QuestionExtractionAnalysisPageRead]
    question_text: str
    content: QuestionExtractionStructuredContentRead | None = None
    answer_options: list[QuestionExtractionAnalysisOptionRead]
    confidence: Decimal = Field(ge=0, le=1)
    needs_review: bool
    corrections: list[QuestionExtractionAnalysisCorrectionRead]
    visual_required: bool


class QuestionExtractionAnalysisBlockRead(StrictQuestionExtractionSchema):
    name: str
    question_count: int


class QuestionExtractionAnalysisRead(StrictQuestionExtractionSchema):
    detected_language: str | None
    total_questions: int
    blocks: list[QuestionExtractionAnalysisBlockRead]
    needs_review_count: int
    corrections_count: int
    visual_required_count: int
    multi_page_question_count: int
    questions: list[QuestionExtractionAnalysisQuestionRead]


class QuestionExtractionAnalysisResultRead(StrictQuestionExtractionSchema):
    run_id: uuid.UUID
    schema_version: int
    processor_name: str
    processor_version: str
    provider_name: str | None
    model_name: str | None
    prompt_version: str | None
    processing_version: str
    analysis: QuestionExtractionAnalysisRead


class QuestionExtractionSuccessfulResultRead(
    StrictQuestionExtractionSchema
):
    run: QuestionExtractionRunRead
    candidate_count: int
    candidates: list[QuestionCandidateRead]
    analysis_result: QuestionExtractionAnalysisResultRead | None = None


class QuestionExtractionOverviewRead(StrictQuestionExtractionSchema):
    source_document_id: uuid.UUID
    media_asset_id: uuid.UUID
    question_source_id: uuid.UUID | None
    uploaded_by_user_id: uuid.UUID | None
    latest_run: QuestionExtractionRunRead | None
    latest_successful_result: QuestionExtractionSuccessfulResultRead | None

File: frontend/src/api/questionExtraction.ts
import { requestJson } from './client'
import type { IsoDateTime, UUID } from './questionEditor'

export type QuestionExtractionRunStatus = 'pending' | 'running' | 'succeeded' | 'failed'
export type QuestionExtractionRunRead = { id: UUID; source_document_id: UUID; run_number: number; status: QuestionExtractionRunStatus; requested_by_user_id: UUID | null; started_at: IsoDateTime | null; completed_at: IsoDateTime | null; failure_message: string | null }
export type QuestionCandidateRead = { id: UUID; sequence_number: number; extracted_text: string; confidence: string | null; source_document_page_id: UUID | null; page_number: number | null }
export type QuestionExtractionAnalysisPageRead = { source_document_page_id: UUID; page_number: number }
export type QuestionExtractionAnalysisOptionRead = { label: string | null; text: string }
export type QuestionExtractionAnalysisCorrectionRead = { original_value: string; normalized_value: string; reason: string }
export type QuestionExtractionAnalysisQuestionRead = { id: UUID; sequence_number: number; question_number: string | null; variant: string | null; source_pages: QuestionExtractionAnalysisPageRead[]; question_text: string; answer_options: QuestionExtractionAnalysisOptionRead[]; confidence: string; needs_review: boolean; corrections: QuestionExtractionAnalysisCorrectionRead[]; visual_required: boolean }
export type QuestionExtractionAnalysisBlockRead = { name: string; question_count: number }
export type QuestionExtractionAnalysisRead = { detected_language: string | null; total_questions: number; blocks: QuestionExtractionAnalysisBlockRead[]; needs_review_count: number; corrections_count: number; visual_required_count: number; multi_page_question_count: number; questions: QuestionExtractionAnalysisQuestionRead[] }
export type QuestionExtractionAnalysisResultRead = { run_id: UUID; schema_version: number; processor_name: string; processor_version: string; provider_name: string | null; model_name: string | null; prompt_version: string | null; processing_version: string; analysis: QuestionExtractionAnalysisRead }
export type QuestionExtractionSuccessfulResultRead = { run: QuestionExtractionRunRead; candidate_count: number; candidates: QuestionCandidateRead[]; analysis_result: QuestionExtractionAnalysisResultRead | null }
export type QuestionExtractionOverviewRead = { source_document_id: UUID; media_asset_id: UUID; question_source_id: UUID | null; uploaded_by_user_id: UUID | null; latest_run: QuestionExtractionRunRead | null; latest_successful_result: QuestionExtractionSuccessfulResultRead | null }

export function getQuestionExtractionOverview(accessToken: string, sourceDocumentId: UUID): Promise<QuestionExtractionOverviewRead> {
  return requestJson<QuestionExtractionOverviewRead>(`/api/v1/sources/${sourceDocumentId}/question-extraction`, { method: 'GET', headers: { Authorization: `Bearer ${accessToken}` } })
}

export function createQuestionExtractionRun(accessToken: string, sourceDocumentId: UUID): Promise<QuestionExtractionRunRead> {
  return requestJson<QuestionExtractionRunRead>(`/api/v1/sources/${sourceDocumentId}/question-extraction/runs`, { method: 'POST', headers: { Authorization: `Bearer ${accessToken}` } })
}
